updater refactoring

mikmaz opened the pull request.

zimnx commented on cmd/updater/logic/updater.go:
```diff
@@ -0,0 +1,134 @@
+		return nil
+	}
+
+	dcRecs := sca.Status.Recommendations.DataCenterRecommendations
+	if dcRecs == nil {
+		return nil
+	}
+
+	return dcRecs
```
```suggestion
	return sca.Status.Recommendations.DataCenterRecommendations
```

zimnx commented on cmd/updater/logic/updater.go:
```diff
@@ -0,0 +1,134 @@
+	logger log.Logger
+}
+
+func NewUpdater(c client.Client, logger log.Logger) (Updater, error) {
```
no need to have error here if it's always nil

zimnx commented on cmd/updater/logic/updater.go:
```diff
@@ -0,0 +1,134 @@
+func getRackRecommendations(dataCenterName string, dcRecs []v1alpha1.DataCenterRecommendations) []v1alpha1.RackRecommendations {
+	for idx := range dcRecs {
+		if dcRecs[idx].Name == dataCenterName {
+			if dcRecs[idx].RackRecommendations == nil {
+				return nil
+			} else {
+				return dcRecs[idx].RackRecommendations
+			}
```
```suggestion
			return dcRecs[idx].RackRecommendations
			
```

zimnx commented on cmd/updater/logic/updater.go:
```diff
@@ -0,0 +1,134 @@
+			Namespace: targetRef.Namespace,
+			Name:      targetRef.Name,
+		}, cluster)
+
```
```suggestion
```

zimnx commented on cmd/updater/logic/updater.go:
```diff
@@ -0,0 +1,134 @@
+		sca := &scas.Items[idx]
+
+		targetRef := sca.Spec.TargetRef
+
```
```suggestion
```

zimnx commented on cmd/updater/logic/updater.go:
```diff
@@ -0,0 +1,134 @@
+
+		dcRecs := getDataCenterRecommendations(sca)
+		if dcRecs == nil {
+			continue
```
log on debug about not found recommendation

zimnx commented on cmd/updater/logic/updater.go:
```diff
@@ -0,0 +1,134 @@
+		}
+
+		dataCenterName := cluster.Spec.Datacenter.Name
+
```
```suggestion
```

zimnx commented on cmd/updater/logic/updater.go:
```diff
@@ -0,0 +1,134 @@
+
+		rackRecs := getRackRecommendations(dataCenterName, dcRecs)
+		if rackRecs == nil {
+			continue
```
log on debug about not found recommendation

zimnx commented on cmd/updater/logic/updater.go:
```diff
@@ -0,0 +1,134 @@
+			rackRec := &rackRecs[j]
+
+			if rackRec.Members == nil {
+				continue
```
log on debug about no members

zimnx commented on cmd/updater/logic/updater.go:
```diff
@@ -0,0 +1,134 @@
+
+			rack := findRack(rackRec.Name, cluster.Spec.Datacenter.Racks)
+			if rack == nil {
+				continue
```
log on debug about not found rack

zimnx commented on cmd/updater/logic/updater.go:
```diff
@@ -0,0 +1,134 @@
+				continue
+			}
+
+			rack.Members = rackRec.Members.Target
```
log on info about applied recommendation

zimnx commented on cmd/updater/logic/updater.go:
```diff
@@ -0,0 +1,134 @@
+		}
+
+		if anyRackUpdated {
+			err = u.client.Update(ctx, cluster)
+			if err != nil {
+				u.logger.Error(ctx, "Cluster update error", "cluster", cluster.Name, "error", err)
+			}
```
```suggestion
			if err := u.client.Update(ctx, cluster); err != nil {
				u.logger.Error(ctx, "Cluster update error", "cluster", cluster.Name, "error", err)
			}
```

zimnx commented on cmd/updater/updater-cmd.go:
```diff
@@ -0,0 +1,54 @@
+			mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
+				Scheme: scheme,
+			})
+
```
```suggestion
```

zimnx commented on cmd/updater/logic/updater.go:
```diff
@@ -0,0 +1,128 @@
+func getDataCenterRecommendations(sca *v1alpha1.ScyllaClusterAutoscaler) []v1alpha1.DataCenterRecommendations {
+	if sca.Status.Recommendations == nil {
+		return nil
+	} else {
```
this doesn't need to be in else block. `if` already prevents from moving forward in the funciton body by having return statement.

zimnx commented on cmd/updater/logic/updater.go:
```diff
@@ -0,0 +1,128 @@
+)
+
+type Updater interface {
+	RunOnce(ctx context.Context)
```
I would expect it can fail so error should be returned back to the caller.

zimnx commented on cmd/updater/logic/updater.go:
```diff
@@ -0,0 +1,128 @@
+	scas := &v1alpha1.ScyllaClusterAutoscalerList{}
+	err := u.client.List(ctx, scas)
+	if err != nil {
+		u.logger.Error(ctx, "fetch SCAs", "error", err)
```
Refactor it into returning errors instead logging it on every error branch. Then you can have a single logging statement outside with returned error

zimnx commented on cmd/updater/logic/updater.go:
```diff
@@ -0,0 +1,128 @@
+			readyMembers := rackStatus.ReadyMembers
+
+			if requestedMembers != readyMembers || requestedMembers == rackRec.Members.Target {
+				continue
```
add log message

zimnx reviewed (commented):
Please add unit tests

zimnx commented on pkg/updater/updater_test.go:
```diff
@@ -0,0 +1,165 @@
+	}{
+		{
+			name: "applied recommendation",
+			rackName: "test-rack-1",
```
It would be much clearer if test input would be pair of recommendation and cluster, and expected state. Right now just by looking at the test input, it's not clear what it verfies.
It is also hard to extend it, because you're running updater outside of individual test.

zimnx reviewed (commented):
Thanks, refactoring of tests is needed.

mikmaz committed:
Updater refactor, unit tests

- Refactored updater
- Changed file structure
- Added filtering off mode SCAs
- Added unit tests

zimnx reviewed (approved):
LGTM, thanks
